File: src/spirite/base/image_data/PaletteSet.java
package spirite.base.image_data;

import spirite.base.brains.PaletteManager;
import spirite.base.brains.PaletteManager.Palette;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

public class PaletteSet {
	public final ImageWorkspace context;
	private final PaletteManager paletteManager;
	
	PaletteSet( ImageWorkspace workspace) {
		this.context = workspace;
		this.paletteManager = workspace.getPaletteManager();

		palettes.add(paletteManager.new Palette("Default"));
		selectedPalette = 0;
	}

	private List<Palette> palettes = new ArrayList<>();
	private int selectedPalette;
	
	public Palette getCurrentPalette() {
		return palettes.get(selectedPalette);
	}
	public void resetPalettes(Collection<Palette> newPalettes) {
		palettes = new ArrayList<>(newPalettes.size());
		palettes.addAll(newPalettes);
		selectedPalette = 0;
	}
	public void addPalette( Palette palette, boolean select) {
		palettes.add(palette);
		if( select)
			selectedPalette = palettes.size()-1;
		paletteManager.triggerPaletteChange();
	}
	public void removePalette( int i) {
		if( i < 0 || i >= palettes.size())
			return;
		
		palettes.remove(i);
		if( selectedPalette >= i)
			selectedPalette--;
		if( palettes.size() == 0) {
			palettes.add(paletteManager.new Palette("Default"));
			selectedPalette = 0;
		}
		paletteManager.triggerPaletteChange();
	}
	public void setSelectedPalette(int i) {
		selectedPalette = i;
		paletteManager.triggerPaletteChange();
	}
	public List<Palette> getPalettes() { return new ArrayList<>(palettes);}
}

File: src/spirite/base/brains/palette/PaletteSet.kt
package spirite.base.brains.palette

import spirite.hybrid.MDebug
import spirite.hybrid.MDebug.WarningType.STRUCTURAL

abstract class PaletteSet {
    abstract val onChangeTrigger: (PaletteSet) -> Unit

    val palettes : List<Palette> get() = _palettes
    private val _palettes = mutableListOf<Palette>()

    var selectedPalette : Int
        get() = _selectedPalette
        set(value) {
            _selectedPalette = value
            onChangeTrigger.invoke(this)
        }
    private var _selectedPalette = 0

    val currentPalette : Palette get() =  _palettes[_selectedPalette]

    fun resetPalettes( newPalette: Collection<Palette>) {
        _palettes.clear()
        _palettes.addAll( newPalette)
        onChangeTrigger.invoke(this)
    }

    fun addPalette( name: String, select: Boolean, raw: ByteArray? = null) : Palette {
        val palette = PSPalette(name, raw)
        _palettes.add(palette)
        if( select)
            _selectedPalette = _palettes.size - 1
        onChangeTrigger.invoke(this)

        return palette
    }
//
//    fun addPalette(palette: Palette, select: Boolean) {
//        _palettes.add( palette)
//        if(select)
//            _selectedPalette = _palettes.size - 1
//        onChangeTrigger.invoke(this)
//    }

    fun removePalette( index: Int) {
        if( index < 0 || index >= _palettes.size) {
            MDebug.handleWarning(STRUCTURAL, "Attempt to remove Palette Out of Bounds ($index not in [0,${_palettes.size})")
            return
        }

        _palettes.removeAt(index)
        if( _selectedPalette >= index)
            _selectedPalette--
        if( _palettes.size == 0) {
            _palettes.add( PSPalette("Default"))
            _selectedPalette = 0
        }
        onChangeTrigger.invoke(this)
    }

    inner class PSPalette( string: String, raw: ByteArray? = null) : Palette(string, raw) {
        override var onChangeTrigger: (Palette) -> Unit = { this@PaletteSet.onChangeTrigger.invoke(this@PaletteSet)}
    }
}